### PR TITLE
fix: Added get_fieldset method to ensure form remains in order

### PR DESCRIPTION
**Pull request by Bernardvdv**
Ensuring the content expiry form field remains in the the same order when compliance number is read-only

### CODE REVIEW

File: tests/test_admin.py [hunk 1]
```diff
@@ -177,6 +177,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
 
         self.assertIn(unpublished_expected_response, decoded_response)
 
+    def test_change_form_fieldset_ordering_draft(self):
+        """
+        The change form should in same order starting with compliance number when in draft
```
Comment (adam-murray) [inline tests/test_admin.py:182]:
```suggestion
        The change form should remain in same order, starting with compliance number when in draft
```

File: tests/test_admin.py [hunk 2]
```diff
@@ -177,6 +177,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
...
+        """
+        content_expiry = PollContentExpiryFactory(version__state=DRAFT)
+        endpoint = self.get_admin_url(ContentExpiry, "change", content_expiry.pk)
+        expexted_fieldset = ['compliance_number', 'created_by', 'version', 'expires']
+        with self.login_user_context(self.get_superuser()):
+            response = self.client.get(endpoint)
+
+        fieldsets = response.context_data['adminform'].fieldsets[0][1]['fields']
+
+        self.assertEqual(expexted_fieldset, fieldsets)
```
Comment (adam-murray) [inline tests/test_admin.py:192]:
```suggestion
        self.assertEqual(expected_fieldset, fieldsets)
```

File: tests/test_admin.py [hunk 3]
```diff
@@ -177,6 +177,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
 
         self.assertIn(unpublished_expected_response, decoded_response)
 
+    def test_change_form_fieldset_ordering_draft(self):
+        """
+        The change form should in same order starting with compliance number when in draft
+        """
+        content_expiry = PollContentExpiryFactory(version__state=DRAFT)
+        endpoint = self.get_admin_url(ContentExpiry, "change", content_expiry.pk)
+        expexted_fieldset = ['compliance_number', 'created_by', 'version', 'expires']
```
Comment (adam-murray) [inline tests/test_admin.py:186]:
```suggestion
        expected_fieldset = ['compliance_number', 'created_by', 'version', 'expires']
```

File: tests/test_admin.py [hunk 4]
```diff
@@ -177,6 +177,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
...
+        with self.login_user_context(self.get_superuser()):
+            response = self.client.get(endpoint)
+
+        fieldsets = response.context_data['adminform'].fieldsets[0][1]['fields']
+
+        self.assertEqual(expexted_fieldset, fieldsets)
+
+    def test_change_form_fieldset_ordering_published(self):
+        """
+        The change form should in same order starting with compliance number when in published
```
Comment (adam-murray) [inline tests/test_admin.py:196]:
```suggestion
        The change form should remain in same order, starting with compliance number when in published
```

File: tests/test_admin.py [hunk 5]
```diff
@@ -177,6 +177,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
...
+        """
+        content_expiry = PollContentExpiryFactory(version__state=PUBLISHED)
+        endpoint = self.get_admin_url(ContentExpiry, "change", content_expiry.pk)
+        expexted_fieldset = ['compliance_number', 'created_by', 'version', 'expires']
```
Comment (adam-murray) [inline tests/test_admin.py:200]:
```suggestion
        expected_fieldset = ['compliance_number', 'created_by', 'version', 'expires']
```

File: djangocms_content_expiry/admin.py [hunk 6]
```diff
@@ -295,6 +295,17 @@ def get_readonly_fields(self, request, obj=None):
...
 
+    def get_fieldsets(self, request, obj=None):
+        """
+        Form field ordering should remain constant when compliance number is read-only
+        :param request: Request object
+        :param obj: Content expiry object
+        :return: Form fieldsets
+        """
+        fieldsets = super(ContentExpiryAdmin, self).get_fieldsets(request, obj)
+        fieldsets[0][1]['fields'] = ['compliance_number', 'created_by', 'version', 'expires']
```
Comment (adam-murray) [inline djangocms_content_expiry/admin.py:306]: Will this require updating every time fields are updated/added to content expiry? If so, should this not use a constant or something similar?

Comment (Bernardvdv) [reply djangocms_content_expiry/admin.py]: Moved fieldset list to a constant

**Review by adam-murray — APPROVED** [review]

File: djangocms_content_expiry/admin.py [hunk 7]
```diff
@@ -295,6 +296,17 @@ def get_readonly_fields(self, request, obj=None):
             return self.readonly_fields + ('compliance_number',)
         return self.readonly_fields
 
+    def get_fieldsets(self, request, obj=None):
```
[the remark this answers is not in the document]
Comment (michaeljcollinsuk) [reply djangocms_content_expiry/admin.py:299]: Although just seen that the constant is used in the tests, so could be a reason to still use it

File: djangocms_content_expiry/admin.py [hunk 8]
```diff
@@ -295,6 +296,17 @@ def get_readonly_fields(self, request, obj=None):
...
+        """
+        Form field ordering should remain constant when compliance number is read-only
+        :param request: Request object
+        :param obj: Content expiry object
+        :return: Form fieldsets
+        """
+        fieldsets = super(ContentExpiryAdmin, self).get_fieldsets(request, obj)
```
Comment (Aiky30) [inline djangocms_content_expiry/admin.py:306]:
```suggestion
        fieldsets = super().get_fieldsets(request, obj)
```

Python 2 logic

File: djangocms_content_expiry/admin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Aiky30) [reply djangocms_content_expiry/admin.py:299]: @Bernardvdv What Michael has stated could be tried which is much easier to read and maintain than relying on the structure returned by get_fieldsets: fieldsets[0][1]['fields'].

[the remark this answers is not in the document]
Comment (Aiky30) [reply djangocms_content_expiry/admin.py:299]: You could still use the constant too if you wanted with that solution: 

```
    fieldsets = (
        (None, {
            'fields': CONTENT_EXPIRY_FIELDSETS,
        })
    )
```

**Review by Aiky30 — CHANGES_REQUESTED** [review]

**Review by michaeljcollinsuk — APPROVED** [review]

File: tests/test_admin.py [hunk 9]
```diff
@@ -177,6 +178,34 @@ def test_change_form_compliance_number_field_is_not_editable_in_unpublished_vers
...
         self.assertIn(unpublished_expected_response, decoded_response)
 
+    def test_change_form_fieldset_ordering_draft(self):
+        """
+        The change form should remain in same order, starting with compliance number when in draft
+        """
+        content_expiry = PollContentExpiryFactory(version__state=DRAFT)
+        endpoint = self.get_admin_url(ContentExpiry, "change", content_expiry.pk)
+        expected_fieldset = CONTENT_EXPIRY_FIELDSETS
+        with self.login_user_context(self.get_superuser()):
```
Comment (Aiky30) [inline tests/test_admin.py:188]:
```suggestion

        with self.login_user_context(self.get_superuser()):
```

Needs a break between the Arrange and Act

**Review by Aiky30 — APPROVED** [review]

**Pull request merged**